sympy/logic/algorithms/z3_wrapper.py

```
@@ -29,54 +75,139 @@ def z3_satisfiable(expr, all_models=False):
+    for var in z3_model:
+        var_name = var.name()
+        # Only process variables that start with 'd' followed by digits
+        if var_name.startswith('d') and var_name[1:].isdigit():
+            var_id = int(var_name[1:])
+            # Only include if this variable ID exists in our reverse encoding
+            if var_id in rev_enc:
+                result[rev_enc[var_id]] = bool(z3_model[var])
+    return result
 
 
 def clause_to_assertion(clause):
+    """Converts a single CNF clause to SMT-LIB assertion format."""
     clause_strings = [f"d{abs(lit)}" if lit > 0 else f"(not d{abs(lit)})" for lit in clause]
     return "(assert (or " + " ".join(clause_strings) + "))"
 
 
-def encoded_cnf_to_z3_solver(enc_cnf, z3):
-    def dummify_bool(pred):
-        return False
-        assert isinstance(pred, AppliedPredicate)
+def find_uninterpreted_functions(expr):
+    uf_arity = {}
+    # Handle bare function heads (symbols that are functions but not called)
+    for head in expr.atoms(UndefinedFunction):
+        uf_arity.setdefault(str(head), None)
+    # Handle function calls like f(a), g(x,y)
+    if AppliedUndef is not None:
+        calls = expr.atoms(AppliedUndef)
+    else:
+        calls = [c for c in expr.atoms(Function) if isinstance(c.func, UndefinedFunction)]
 
-        if pred.function in [Q.positive, Q.negative, Q.zero]:
-            return pred
-        else:
-            return False
+    for call in calls:
+        name = str(call.func)
+        arity = len(call.args)
+        prev = uf_arity.get(name)
+        uf_arity[name] = arity if prev is None else max(prev or 0, arity)
 
-    s = z3.Solver()
+    return uf_arity
 
-    declarations = [f"(declare-const d{var} Bool)" for var in enc_cnf.variables]
-    assertions = [clause_to_assertion(clause) for clause in enc_cnf.data]
 
-    symbols = set()
+def collect_all_uninterpreted_functions(enc_cnf):
+    """Returns a dictionary, mapping uninterpreted function names to their maximum arity."""
```

They shouldn't have multiple possible arities. If that happens you should probably raise an exception.